File: Go/GerberWriter.go
package main

import (
	"fmt"
	"image"
	"math"
	"os"
	"strings"
	"time"
)

func FinishFile(file *os.File, buff []byte) {
	buff = append(buff, []byte("M02*")...)
	file.Write(buff)
	file.Close()
}

func NumRepr(num float64) string {
	integer := int(num)
	fraction := int(math.Mod(num, 1.0))
	return fmt.Sprintf("%04d%06d", integer, fraction)
	// return strings.ReplaceAll(fmt.Sprintf("%011.6f", num), ".", "")
}

func WriteHeader(buff []byte, gerberType string) []byte {
	buff = append(buff, []byte("%%TF.GenerationSoftware,ImageToGerber,Design,1.1.0*%\n")...)
	buff = append(buff, []byte("%%TF.SameCoordinates,Original*%\n")...)
	if gerberType == "F_Cu" {
		// Front Copper
		buff = append(buff, []byte("%TF.FileFunction,Copper,L1,Top*%\n")...)
		buff = append(buff, []byte("%TF.FilePolarity,Positive*%\n")...)
	} else if gerberType == "B_Cu" {
		// Back Copper
		buff = append(buff, []byte("%TF.FileFunction,Copper,L2,Bot*%\n")...)
		buff = append(buff, []byte("%TF.FilePolarity,Positive*%\n")...)
	} else if gerberType == "F_Mask" {
		// Front Solder Mask
		buff = append(buff, []byte("%TF.FileFunction,Soldermask,Top*%\n")...)
		buff = append(buff, []byte("%TF.FilePolarity,Negative*%\n")...)
	} else if gerberType == "B_Mask" {
		// Back Solder Mask
		buff = append(buff, []byte("%TF.FileFunction,Soldermask,Bot*%\n")...)
		buff = append(buff, []byte("%TF.FilePolarity,Negative*%\n")...)
	} else if gerberType == "F_SilkS" {
		// Front Silk screen
		buff = append(buff, []byte("%TF.FileFunction,Legend,Top*%\n")...)
		buff = append(buff, []byte("%TF.FilePolarity,Positive*%\n")...)
	} else if gerberType == "B_SilkS" {
		// Back Silk screen
		buff = append(buff, []byte("%TF.FileFunction,Legend,Bot*%\n")...)
		buff = append(buff, []byte("%TF.FilePolarity,Positive*%\n")...)
	} else if gerberType == "Edge_Cuts" {
		// Edge Cuts
		buff = append(buff, []byte("%%TF.FileFunction,Profile,NP*%\n")...)
	}

	buff = append(buff, []byte("%FSLAX46Y46*%\n")...) //Number Format
	buff = append(buff, []byte("%MOMM*%\n")...)       //Millimeters
	buff = append(buff, []byte("%LPD*%\n")...)        //Dark polarity
	buff = append(buff, []byte("G04 == End Header ==*\n")...)

	return buff
}

func GerberCreateLine(buff []byte, scaleWidth float64, scaleHeight float64, img *image.Image, x int, y int, rectsize int) []byte {

	buff = append(buff, []byte(fmt.Sprintf("D%d*\n", rectsize+10))...) //Use Rectangle Object
	buff = append(buff, []byte(
		fmt.Sprintf("X%sY%sD03*\n",
			NumRepr(scaleWidth*(float64(x)-float64(rectsize)/2)),
			NumRepr(scaleHeight*(float64((*img).Bounds().Dy()-y)+1.0/2)),
		),
	)...) //Place rectangle

	return buff
}

func GenerateGerberFillLines(img image.Image, gerberWidth float64, gerberHeight float64, filename string, gerberType string, printer Printer) {
	// file = CreateFile(filename, gerberType)
	file := CreateFile(filename + "-" + gerberType + ".gbr")

	scaleWidth := gerberWidth / float64(img.Bounds().Dx())
	scaleHeight := gerberHeight / float64(img.Bounds().Dy())

	printer.Print("Writing Gerber")
	buff := make([]byte, 0)
	buff = WriteHeader(buff, gerberType)

	for i := 1; i <= img.Bounds().Dx(); i++ {
		buff = append(buff, []byte(fmt.Sprintf("%%ADD%dR,%.6fX%.6f*%%\n", i+10, scaleWidth*float64(i), scaleHeight))...) //Rectangle Object
	}

	tStart := time.Now()
	tLast := time.Time{}
	rectsize := 0
	for y := 0; y < img.Bounds().Dy(); y++ {
		for x := 0; x < img.Bounds().Dx(); x++ {
			//Get pixel value at this point
			r, _, _, _ := img.At(x, y).RGBA()
			if r == 65535 {
				rectsize += 1
				// Handle line
			} else if rectsize > 0 {
				buff = GerberCreateLine(buff, scaleWidth, scaleHeight, &img, x, y, rectsize)
				// DrawLine(x, y, rectsize)
				rectsize = 0
			}
		}
		// Handle last run
		if rectsize > 0 {
			buff = GerberCreateLine(buff, scaleWidth, scaleHeight, &img, img.Bounds().Dx()-1, y, rectsize)
			// DrawLine(img.Bounds().Dx()-1, y, rectsize)
			rectsize = 0
		}
		tNow := time.Now()

		if tNow.Sub(tLast) > 100*time.Millisecond {
			progress := 10 * y / img.Bounds().Dy()
			bar := strings.Repeat("=", progress) + strings.Repeat(" ", 10-progress)
			printer.Print(fmt.Sprintf("Writing Gerber %s [%s] Time:%v", , bar, tNow.Sub(tStart)))
			tLast = tNow
		}
	}
	// fmt.Println("Time:", t2.Sub(t1))

	FinishFile(file, buff)
}

func GenerateGerberTrace(img image.Image, gerberWidth float64, gerberHeight float64, filename string, gerberType string) {
	segments := LineDetection(img)
	_ = segments

	scaleWidth := gerberWidth / float64(img.Bounds().Dx())
	scaleHeight := gerberHeight / float64(img.Bounds().Dy())

	file := CreateFile(filename + "-" + gerberType + ".gbr")

	fmt.Println(TERM_BLUE + "== Writing Gerber ==" + TERM_RESET)
	buff := make([]byte, 0)
	buff = WriteHeader(buff, gerberType)

	buff = append(buff, []byte("G01*\n")...) //Linear interpolation mode
	// # file.write("%TA.AperFunction,Profile*%")
	buff = append(buff, []byte("%ADD10C,0.200000*%")...) //Circle aperture
	// # file.write("%TD*%")
	buff = append(buff, []byte("D10*")...) //Use aperture

	for _, segment := range segments {
		buff = append(buff, []byte(
			fmt.Sprintf("X%sY%sD02*\n",
				NumRepr(scaleWidth*segment.P1.X),
				NumRepr(scaleHeight*(float64(img.Bounds().Dy())-segment.P1.Y)),
			),
		)...)
		buff = append(buff, []byte(
			fmt.Sprintf("X%sY%sD02*\n",
				NumRepr(scaleWidth*segment.P2.X),
				NumRepr(scaleHeight*(float64(img.Bounds().Dy())-segment.P2.Y)),
			),
		)...)
	}

	FinishFile(file, buff)
	fmt.Println(TERM_GREY + "== Done Writing Gerber ==" + TERM_RESET)
}

File: Python/GerberWriter.py

#https://www.ucamco.com/en/guest/downloads/gerber-format

from Global import *
import numpy as np

def CreateFile(filename, gerberType):
    filename+=F"-{gerberType}.gbr"
    try:
        return open(filename, "w")
    except:
        Error(F"Failed to create file \"{filename}\"")
        return None

def FinishFile(file):
    file.write("M02*")
    file.close()

def NumRepr(num):
    integer = int(num)
    fraction = num
    return F"{num:011.6f}".replace(".","")

def WriteHeader(file, gerberType):
    file.write("%TF.GenerationSoftware,ImageToGerber,Design,1.1.0*%\n")
    file.write("%TF.SameCoordinates,Original*%\n")
    #Copper
    if gerberType=="F_Cu":
        file.write("%TF.FileFunction,Copper,L1,Top*%\n")
        file.write("%TF.FilePolarity,Positive*%\n")
    elif gerberType=="B_Cu":
        file.write("%TF.FileFunction,Copper,L2,Bot*%\n")
        file.write("%TF.FilePolarity,Positive*%\n")
    #Solder Mask
    elif gerberType=="F_Mask":
        file.write("%TF.FileFunction,Soldermask,Top*%\n")
        file.write("%TF.FilePolarity,Negative*%\n")
    elif gerberType=="B_Mask":
        file.write("%TF.FileFunction,Soldermask,Bot*%\n")
        file.write("%TF.FilePolarity,Negative*%\n")
    #Silk screen
    elif gerberType=="F_SilkS":
        file.write("%TF.FileFunction,Legend,Top*%\n")
        file.write("%TF.FilePolarity,Positive*%\n")
    elif gerberType=="B_SilkS":
        file.write("%TF.FileFunction,Legend,Bot*%\n")
        file.write("%TF.FilePolarity,Positive*%\n")
    #Edge Cuts
    elif gerberType=="Edge_Cuts":
        file.write("%TF.FileFunction,Profile,NP*%\n")

    file.write("%FSLAX46Y46*%\n") #Number Format
    file.write("%MOMM*%\n") #Millimeters
    file.write("%LPD*%\n") #Dark polarity
    file.write("G04 == End Header ==*\n")


def GeneratePixelated(img, filename):
    file = CreateFile(filename)
    WriteHeader(file)
    file.write("%ADD10R,1X1*%\n") #Rectangle Object
    file.write("D10*\n") #Use Rectangle Object

    height, width = img.shape

    print("> Writing Gerber")
    for y in range(height):
        for x in range(width):
            if img[y][x]:
                file.write(F"X{NumRepr(x)}Y{NumRepr(height-y)}D03*\n")
        ProgressBar(y, 0, height-1)
    print()

def GeneratePixelatedFillLines(img, gerberDim, filename, gerberType):
    file = CreateFile(filename, gerberType)
    WriteHeader(file, gerberType)
    height, width = img.shape
    scaleWidth = gerberDim.Width/width
    scaleHeight = gerberDim.Height/height

    lines = 0
    rectsize = 0

    def DrawLine(x, rectsize, line):
        file.write(F"%ADD{line+10}R,{scaleWidth*rectsize:.6f}X{scaleHeight:.6f}*%\n") #Rectangle Object
        file.write(F"D{line+10}*\n") #Use Rectangle Object
        file.write(F"X{NumRepr(scaleWidth*(x-rectsize/2))}Y{NumRepr(scaleHeight*(height-y+1/2))}D03*\n") #Place rectangle

    print("> Writing Gerber")
    ts=time.time()
    for y in range(height):
        for x in range(width):
            if img[y][x]:
                rectsize += 1
            #Handle line
            elif rectsize>0:
                DrawLine(x, rectsize, lines)
                rectsize = 0
                lines+=1
        #Handle last line
        if rectsize>0:
            DrawLine(width-1, rectsize, lines)
            rectsize = 0
            lines+=1
        ProgressBar(y, 0, height-1)
        TimeDisplay(ts)

    print()
    FinishFile(file)

def GeneratePixelatedOctree(octree, filename, dim):
    file = CreateFile(filename)
    WriteHeader(file)

    sizes = 0
    depths = {}

    for item in octree:
        if not item[0] in depths:
            depths |= {item[0]:10+sizes}
            scale = min(dim)/(2**item[0])
            file.write(F"%ADD{10+sizes}R,{scale:.6f}X{scale:.6f}*%\n") #Rectangle Object
            sizes+=1

    for item in octree:
        file.write(F"D{depths[item[0]]}*\n") #Use Rectangle Object
        s=min(dim)/(2**item[0])
        x=item[1] + s/2
        y=item[2] + s/2
        file.write(F"X{NumRepr(x)}Y{NumRepr(dim[1]-y)}D03*\n")


def GenerateTraceFromLoops(lineloops, dim, filename, gerberType):
    file = CreateFile(filename, gerberType)
    WriteHeader(file, gerberType)

    file.write("G01*\n") #Linear interpolation mode
    # file.write("%TA.AperFunction,Profile*%")
    file.write("%ADD10C,0.200000*%") #Circle aperture
    # file.write("%TD*%")
    file.write("D10*") #Use aperture

    height, width = dim

    for loop in lineloops:
        file.write(F"X{NumRepr(loop[0][0])}Y{NumRepr(height-loop[0][1])}D02*\n")
        for point in loop[1:]:
            file.write(F"X{NumRepr(point[0])}Y{NumRepr(height-point[1])}D01*\n")
    print()
    FinishFile(file)

def GenerateTraceFromSegments(segments, imgDim, gerberDim, filename, gerberType):
    file = CreateFile(filename, gerberType)
    WriteHeader(file, gerberType)

    scaleWidth = gerberDim.Width/imgDim.Width
    scaleHeight = gerberDim.Height/imgDim.Height

    file.write("G01*\n") #Linear interpolation mode
    # file.write("%TA.AperFunction,Profile*%")
    file.write("%ADD10C,0.200000*%") #Circle aperture
    # file.write("%TD*%")
    file.write("D10*") #Use aperture

    print("> Writing Gerber")
    ts=time.time()
    for i, line in enumerate(segments):
        file.write(F"X{NumRepr(scaleWidth*line[0][0])}Y{NumRepr(scaleHeight*(imgDim.Height-line[0][1]))}D02*\n")
        file.write(F"X{NumRepr(scaleWidth*line[1][0])}Y{NumRepr(scaleHeight*(imgDim.Height-line[1][1]))}D01*\n")
        ProgressBar(i, 0, len(segments)-1)
        TimeDisplay(ts)
    print()

    FinishFile(file)


